File: designPatterns/SOLID_design_principles/InterfaceSegregationPrinciple/typescript/passInterfaceSegregationPrinciple.ts
interface Entity2 {
  name: string;
}

interface Mover {
  move(): void;
}

interface Attacker {
  attackDamage: number;
  attack(targetEntity: any): void;
}

interface HasHealth {
  health: number;
  takeDamage(amount: number): void;
}

class Character2 implements Entity2, Mover, Attacker, HasHealth {
  name: string;
  attackDamage: number;
  health: number;

  constructor(name: string, attackDamage: number, health: number) {
    this.name = name;
    this.attackDamage = attackDamage;
    this.health = health;
  }

  move(): void {
    console.log(`${this.name} moved`);
  }

  attack(targetEntity: any): void {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    );
    targetEntity.takeDamage(this.attackDamage);
  }

  takeDamage(amount: number): void {
    this.health -= amount;
    console.log(`${this.name} has ${this.health} health remaining`);
  }
}

class Wall2 implements Entity2, HasHealth {
  name: string;
  health: number;

  constructor(name: string, health: number) {
    this.name = name;
    this.health = health;
  }

  takeDamage(amount: number): void {
    this.health -= amount;
    console.log(`${this.name} has ${this.health} health remaining`);
  }
}

class Turret2 implements Entity2, Attacker {
  name: string;
  attackDamage: number;

  constructor(name: string, attackDamage: number) {
    this.name = name;
    this.attackDamage = attackDamage;
  }

  attack(targetEntity: any): void {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    );
    targetEntity.takeDamage(this.attackDamage);
  }
}

const turret2 = new Turret2("Turret", 5);
const character2 = new Character2("Character", 3, 100);
const wall2 = new Wall2("Wall", 200);

turret2.attack(character2);
character2.move();
character2.attack(wall2);

File: designPatterns/SOLID_design_principles/InterfaceSegregationPrinciple/python/passInterfaceSegregationPrinciple.py
from abc import ABC, abstractmethod


class Entity(ABC):
    def __init__(self, name):
        self.name = name


class Mover(ABC):
    @abstractmethod
    def move(self):
        pass


class Attacker(ABC):
    def __init__(self, attack_damage):
        self.attack_damage = attack_damage

    @abstractmethod
    def attack(self, target_entity):
        pass


class HasHealth(ABC):
    def __init__(self, health):
        self.health = health

    @abstractmethod
    def take_damage(self, amount):
        pass


class Character(Entity, Mover, Attacker, HasHealth):
    def __init__(self, name, attack_damage, health):
        Entity.__init__(self, name)
        Attacker.__init__(self, attack_damage)
        HasHealth.__init__(self, health)

    def move(self):
        print(f"{self.name} moved")

    def attack(self, target_entity):
        print(
            f"{self.name} attacked {target_entity.name} for {self.attack_damage} damage",
        )
        target_entity.take_damage(self.attack_damage)

    def take_damage(self, amount):
        self.health -= amount
        print(f"{self.name} has {self.health} health remaining")


class Wall(Entity, HasHealth):
    def __init__(self, name, health):
        Entity.__init__(self, name)
        HasHealth.__init__(self, health)

    def take_damage(self, amount):
        self.health -= amount
        print(f"{self.name} has {self.health} health remaining")


class Turret(Entity, Attacker):
    def __init__(self, name, attack_damage):
        Entity.__init__(self, name)
        Attacker.__init__(self, attack_damage)

    def attack(self, target_entity):
        print(
            f"{self.name} attacked {target_entity.name} for {self.attack_damage} damage",
        )
        target_entity.take_damage(self.attack_damage)


turret = Turret("Turret", 5)
character = Character("Character", 3, 100)
wall = Wall("Wall", 200)

turret.attack(character)
character.move()
character.attack(wall)
